app/controllers/api/v2/courses_controller.rb

```diff
@@ -49,9 +51,11 @@ def publishable
       end
 
       def update
+        study_mode_updated = params.fetch(:course, {})[:study_mode] != @course.study_mode
```

Comment:
Does it make more sense to move this check to be a guard statement within https://github.com/DFE-Digital/manage-courses-backend/pull/709/files#diff-7d6b888ba704f4684c5b576a1dd07676R104 

Also, there's duplication, should use this: https://github.com/DFE-Digital/manage-courses-backend/pull/709/files#diff-7d6b888ba704f4684c5b576a1dd07676R197-R199

Reply:
Have used the study_mode_params method. The issue around moving this into the update_site_status method as a  guard clause is that it by that time, the course object will have already been updated. So the course.study_mode will always be the same as the params study mode.

Reply:
Have had a chat with @davidgisbey IRL, now I understand why the `study_mode_updated` value needs to be cached before the `update_course` method. We've thought about a few alternatives to the current implementation.

First idea is something like this (pseudocode)

```diff
def update
  update_course
  update_enrichment
  update_sites
+ update_site_statuses
  should_sync = site_ids.present? && @course.recruitment_cycle.current?
  has_synced? if should_sync

        
...

def update_course
  return unless course_params.values.any?
  return unless @course.course_params_assignable(course_params)
+ course_params_without_study_mode = course_params
+ @course.assign_attributes(course_params_without_study_mode)
  @course.save
end

...

def update_site_statuses
+ guard_clause_goes_here
+ @course.study_mode = study_mode_params
  @course.site_statuses.each do |site_status|
    if site_status.vac_status != 'no_vacancies'
      update_vac_status(@course.study_mode, site_status)
    end
  end
end
```

Basically, we update the study mode separately from the other attributes as part of updating site statuses.

Second idea:

```diff
def update_site_statuses
+ guard_clause_goes_here
+ @course.study_mode = study_mode_params
  @course.site_statuses.each do |site_status|
    if site_status.vac_status != 'no_vacancies'
      update_vac_status(@course.study_mode, site_status)
    end
  end
end

...

def course_params
  params
    .fetch(:course, {})
      [snip]
      :start_date,
      :applications_open_from,
-     :study_mode,
    )
end
```

Which is removing the `study_mode` from the regular params and treating it separately.

There's also a third idea around checking if the `@course.has_changed` or something similar, must be some methods we can call in ActiveModel.

But to be honest, for my own 2 cents, I think the current implementation is actually fine now that I understand why it is the way it is.

Reply:
I had a problem around this too. My idea/suggestion is to rename `update_site_statuses` to `update_site_status_vac_statuses`, and use an if statement in `update` to communicate it's intent more clearly:

```
      if @course.study_mode_previously_changed?
        update_site_status_vac_statuses
      end
```

https://github.com/DFE-Digital/manage-courses-backend/pull/709#discussion_r313833165